### PR TITLE
Refactor Walkontable Selection rendering module 

### PR DESCRIPTION
**Pull request by budnix**
### Context

The PR refactors and extends (with some braking change) the Walkontable Selection rendering module for the ability to highlight the table's headers. The work presented here will be the basis for the navigation in headers ([#1062](https://github.com/handsontable/dev-handsontable/issues/1062)).

The changes include:
 * Extract the drawing logic from the _Selection_ class to the separate parent class called _SelectionManager_. This allowed simplifying the _Selection_ class (used by the Handsontable through the _Highlight_ class) to a thing that only holds coordinates, border settings, and CSS class names to render be rendered in the correct cell.
 * Extract and divide into smaller loops the main selection loop that was used to scan intersected elements;
 * Create the Selection module by moving the related files and code, such as the _Selection_ class, its manager, _Border_ and selection-related methods from _Table_ class to the "selection" directory;
 * Extend the overlay's Table instances for the method that allows getting the number of the rendered headers;
 * Reorganize the naming and adjust the Handsontable Selection module to support the new Walkontable Selection API;
 * Implement ASCIITable test matcher for the Walkontable tests for writing more convenient visual tests for selections;

#### How to test it

Currently, it's hard to test the feature visually as the Walkontable only renders coordinates that are controlled by the Handsontable Selection module. That module does not allow navigating cell selection (focus) in headers. However, it can be bypassed by changing the coords directly in the object and calling the `render()` method.

Setup the selection by clicking any cell, execute:
```js
hot.selection.highlight.focus.cellRange.from.row = -2
hot.selection.highlight.focus.cellRange.from.col = -1
hot.selection.highlight.focus.cellRange.to.row = -2
hot.selection.highlight.focus.cellRange.to.col = -1
hot.render();
```  
results:
image

#### Breaking changes
 * Remove the `classNames` key from the hook `beforeHighlightingColumnHeader` and `beforeHighlightingRowHeader` options ([fun fact, the documentation doesn't say anything about that key](https://handsontable.com/docs/javascript-data-grid/api/hooks/#beforehighlightingrowheader)); 

#### Minor changes
 * Add two new methods to the public API of the `CellCoord` class (`isHeader` and `isCell`);
 * Add two new methods to the Core (`countRowHeaders` and `countColHeaders `);
 * Mark the `beforeRemoveCellClassNames` hook as deprecated. The _MergeCells_ plugin used it, but after refactoring, it's not necessary anymore;

_The PR is a part of the accessibility project._

### How has this been tested?

I tested the changes locally, and I covered the feature with new E2E and unit tests.

### Types of changes

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Related issue(s):
1. part of https://github.com/handsontable/dev-handsontable/issues/1062

### Affected project(s):
- [x] `handsontable`

### Checklist:

- [x] I have reviewed the guidelines about [Contributing to Handsontable](https://github.com/handsontable/handsontable/blob/master/CONTRIBUTING.md) and I confirm that my code follows the code style of this project.
- [x] I have signed the [Contributor License Agreement](https://docs.google.com/forms/d/e/1FAIpQLScpMq4swMelvw3-onxC8Jl29m0fVp5hpf7d1yQVklqVjGjWGA/viewform?c=0&w=1)
- [x] My change requires a change to the documentation.

### CODE REVIEW

File: handsontable/src/3rdparty/walkontable/src/selection/selection.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
+import { mixin } from '../../../../helpers/object';
+import localHooks from '../../../../mixins/localHooks';
+
+/**
+ * The Selection class allows highlighting (by applying CSS class) the table's cells or headers
```
Comment (jansiegel) [inline handsontable/src/3rdparty/walkontable/src/selection/selection.js:5]: Is this description correct? If I'm reading this class correctly, it doesn't apply any CSS classes; this seems to be done in the selection manager.

**Review by jansiegel — APPROVED** [review]
I'm not 100% sure what's the desired result of this PR.
I went through all the changes and apart from the JSDoc I'm not sure about everything looks fine + the example provided in the PR description does select the header, but, for example, `hot.getSelected()` still returns the previously-selected coordinates, not the header coordinates.
I'm assuming things like that will be resolved in future PRs?

File: handsontable/src/3rdparty/walkontable/src/selection/selection.js (hunk 1 above)
Comment (budnix) [reply handsontable/src/3rdparty/walkontable/src/selection/selection.js:5]: Generally, it's true. The Selection class as a protected API is used by Handsontable to control where (based on the coords) the CSS classes (passed in constructor `new Selection(cellRange, { className: 'my-css-class' })`) should be applied. From the side of Handsontable as an API consumer, it seems to make sense 🤔

**Comment by budnix** [comment]
> I went through all the changes and apart from the JSDoc I'm not sure about everything looks fine + the example provided in the PR description does select the header, but, for example, `hot.getSelected()` still returns the previously-selected coordinates, not the header coordinates.
> I'm assuming things like that will be resolved in future PRs?

True, the `hot.getSelected()` is a part of the Handsontable AP. The changes that fix other Handsontable-like things will be added in a separate PR.

**Pull request merged**